src/ts/index.ts

```diff
@@ -0,0 +1,296 @@
+  public signEncodeOrder(
+    owner: Wallet,
+    order: Order,
+    executedAmount: BigNumberish,
+  ): void {
+    if (!this.domainSeparator) {
+      throw new Error("domain separator not specified");
+    }
+    const signature = signOrder(owner, this.domainSeparator, order);
+    this.encodeOrder(order, executedAmount, signature);
+  }
+
+  private tokenIndex(token: string): number {
+    // NOTE: Verify and normalize the address into a case-checksummed address.
+    // Not only does this ensure validity of the addresses early on, it also
+    // makes it so `0xff...f` and `0xFF..F` map to the same ID.
+    const tokenAddress = ethers.utils.getAddress(token);
+
+    let tokenIndex = this._tokenMap[tokenAddress];
+    if (tokenIndex === undefined) {
+      tokenIndex = this._tokens.length;
+      this._tokens.push(tokenAddress);
+      this._tokenMap[tokenAddress] = tokenIndex;
+    }
+
+    return tokenIndex;
+  }
+}
+
+/**
+ * Compute the 32-byte digest for the specified order.
+ * @param domainSeparator The domain separator to add to the digest. This is
+ * used by the smart contract to ensure order's can't be replayed across
+ * different applications (domains), but also different deployments (as the
+ * contract chain ID and address is used to salt the domain separator value).
+ * @param order The order to compute the digest for.
+ * @return Hex-encoded 32-byte order digest.
+ */
+export function hashOrder(domainSeparator: string, order: Order): string {
+  return ethers.utils.keccak256(
+    ethers.utils.defaultAbiCoder.encode(
+      [
+        "bytes32",
+        "address",
+        "address",
+        "uint256",
+        "uint256",
+        "uint32",
+        "uint32",
+        "uint256",
+        "uint8",
+        "bool",
+      ],
+      [
+        domainSeparator,
+        order.sellToken,
+        order.buyToken,
+        order.sellAmount,
+        order.buyAmount,
+        order.validTo,
+        order.nonce,
+        order.tip,
+        order.kind,
+        order.partiallyFillable,
+      ],
+    ),
+  );
+}
+
+/**
+ * Returns the signature for the specified order.
+ * @param owner The owner for the order used to sign.
+ * @param domainSeparator The domain separator to add to the digest used for
+ * signing. See {@link hashOrder} for more details.
+ * @param order The order to sign.
+ * @return Signature for the order.
+ */
+export function signOrder(
+  owner: Wallet,
+  domainSeparator: string,
+  order: Order,
+): Signature {
+  const digest = hashOrder(domainSeparator, order);
+  // NOTE: We need to sign directly with the signing key here on an Ethers.js
+  // wallet and not the `Signer` API, as the latter uses the `"\x19Ethereum
+  // Signed Message:\n" + len(message)` prefix. This is likely to change in the
+  // future as wallets generally don't provide a method for doing this as it is
+  // considered unsafe.
+  // <https://github.com/gnosis/oba-contracts/issues/129>
+  return owner._signingKey().signDigest(digest);
```

Comment:
Will this work as well when used from a frontend with metamask?

Reply:
No :( see #129 and #131

Reply:
I plan on adding this next.